File: 0518-coin-change-ii/0518-coin-change-ii.java
class Solution {
    public int change(int amount, int[] coins) {
        int[][] dp = new int[amount + 1][coins.length + 1];
        for(int i = 0; i < dp.length; ++i)
            for(int j = 0; j < dp[i].length; ++j)
                dp[i][j] = -1;
        
        return sum(amount, amount, coins, 0, dp);
    }
    
      public static int sum(int amount, int target, int[] coins, int j, int[][] dp){
          if(target < 0)
              return 0;
          if(target == 0)
              return 1;
          if(dp[target][j] != -1)
              return dp[target][j];
          int ans = 0;
          for(int i = j; i < coins.length; ++i){
              int a = sum(amount, target - coins[i], coins, i, dp);
              ans += a;
          }
          dp[target][j] = ans;
          return ans;
      }
}

File: 0518-coin-change-ii/0518-coin-change-ii.cpp
class Solution {
public:
    int change(int amount, vector<int>& coins) {
        int n = coins.size();
        vector<vector<int>> dp(n, vector<int>(amount + 1, -1));
        return solve(coins, amount, 0, dp);
    }
    
    int solve(vector<int>& coins, int amount, int k, vector<vector<int>>& dp){
        if(amount < 0)
            return 0;
        else if(amount == 0)
            return 1;
        if(dp[k][amount] != -1)
            return dp[k][amount];
        int ans = 0;
        for(int i = k; i < coins.size(); ++i)
            ans += solve(coins, amount - coins[i], i, dp);
        
        return dp[k][amount] = ans;
            
    }
};